[Lim Jin Yin] iP

miljyy opened the pull request:
## Task Buddy
> “Do the hard jobs first. The easy jobs will take care of themselves.” – Dale Carnegie ([source](https://www.goodreads.com/quotes/650081-do-the-hard-jobs-first-the-easy-jobs-will-take))

**Task Buddy** frees your mind of having to remember things you need to do. It's:

- text-based
- easy to learn
- ~~FAST~~ **_SUPER_ FAST** to use

### All you need to do is:

1. download it from [here](https://github.com/nus-cs2103-AY2324S1/ip/pull/305#).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

### Features:

 - Managing tasks
 - Managing deadlines (coming soon)
 - Reminders (coming soon)

If you are a Java programmer, you can also use it to practice Java. Here's the `main` method:
``` ruby
public static void main(String[] args) {
    new Buddy().run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

miljyy committed:
Created new chatbot in Buddy.java with simple greeting and exit, deleted Duke.java

miljyy committed:
Improved Buddy to echo user commands and exit on "bye"

miljyy committed:
Added ability to add and store user input using a list.

miljyy committed:
Create Task.java class

miljyy committed:
Add ability to mark tasks as done and not done

miljyy committed:
Create Deadline, Event, Todo classes

miljyy committed:
Add support for tracking three types of tasks: todos, deadlines and events

miljyy committed:
Implement semi-automated testing using the I/O redirection technique

miljyy committed:
Handle exceptions in main.

miljyy committed:
Add support for deleting tasks from the list.

miljyy committed:
Create tasks.txt in data folder

miljyy committed:
Add TaskList.java and edit Buddy.java to use TaskList class

miljyy committed:
Add TaskType.java enum

miljyy committed:
Update TaskType enum

miljyy committed:
Update Buddy tasks to work with TaskList

miljyy committed:
Add delete function

miljyy committed:
Update Buddy to save tasks to file

miljyy committed:
Update Buddy to load previous tasks from data file

miljyy committed:
Merge branch 'branch-Level-7'

miljyy committed:
Create parser with local date parser

miljyy committed:
Update Buddy to understand dates

miljyy committed:
Merge branch 'branch-Level-8'

miljyy committed:
Organise classes into buddy package

miljyy committed:
Update storage and Ui classes

miljyy committed:
Create enum for command types

miljyy committed:
Update storage and Ui utility classes

inezkok commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,37 @@
+    protected String description;
+    protected boolean isDone;
+
+    public Task(String description, boolean isDone) {
```
I think that the constructor for a Task doesn't need to have the isDone argument since when a Task is first created it should be initialised as not done

inezkok commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,188 @@
+import java.io.*;
```
By the coding standard, import statements should avoid using * so maybe you can split it up into the individual import statements needed

inezkok commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,188 @@
+
+    public void deleteTask(int taskIndex) throws BuddyException {
+        if (tasks.size() == 0 ) {
+            System.out.println("There are no tasks in your list:");
```
I think instead of printing the error statement, you can throw a DukeException with the error message

inezkok commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,188 @@
+            Task deletedTask = tasks.remove(taskIndex);
+            System.out.println("Noted. I've removed this task:");
+            System.out.println(deletedTask.toString());
+            if (tasks.size() == 1) {
```
You don't need to specifically identify if the size is equal to 1 and instead can combine the two cases into one since in both cases they print out the same statement

inezkok commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,188 @@
+        }
+    }
+
+    public void processCommand(String command) throws BuddyException {
```
I think this method should be in the Parser class that is in charge of making sense of the user's input

inezkok commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,188 @@
+        }
+    }
+
+    public void saveTasks() {
```
I think he save and load tasks methods should be in the Storage class

inezkok reviewed (commented):
Overall, it looks good maybe just some more OOP (shift some of the methods in tasks to other classes)

matochichap commented on src/main/java/Buddy.java:
```diff
@@ -0,0 +1,59 @@
+                Integer taskIndex = Integer.valueOf(arrOfCmd[1]) - 1;
+
+                try {
+                    // Task thisTask = tasks.getTask(taskIndex);
```
Maybe this comment could be removed?

matochichap commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+public class Deadline extends Task {
+    protected LocalDate by;
+
+    public Deadline(String description, String by, boolean isDone) {
```
Perhaps some documentation could be added for the public methods?

matochichap commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,37 @@
+public class Task {
```
I love the use of polymorphism here!

matochichap commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,188 @@
+        boolean isDone = parts[1].equals("1");
+        String description = parts[2];
+
+        switch (taskType) {
```
Maybe the indentation level for the cases could be 4 spaces less?

matochichap commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,188 @@
+import java.io.*;
+import java.util.ArrayList;
+
+public class TaskList {
```
Perhaps deleteTask, markAsDone and markAsNotDone could return something so that it would be easier to do tests on the class later on and printing of statements could be handled by another class

matochichap commented on src/main/java/TaskType.java:
```diff
@@ -0,0 +1,3 @@
+public enum TaskType {
+    TODO, DEADLINE, EVENT
```
I love this idea of using enums here! I didn't think of using it to differentiate task type in my code

matochichap reviewed (commented):
The code looks great! I really liked some of the ideas you had like with using enums!

miljyy committed:
Fix bugs in storage class

miljyy committed:
Update Storage class

ivyy-poison commented on src/main/java/Buddy.java:
```diff
@@ -0,0 +1,59 @@
+        String greeting = String.format("Hello! I'm %s\n", name);
+        String inquiry = "What would you like to do?\n";
+        String exit = "Bye. Hope to see you again soon!\n";
+
```
I like the naming scheme used as it makes your code very readable!

ivyy-poison commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,40 @@
+        return "E";
+    }
+
+    @Override
+    public String toSaveFileFormat() {
+        return String.format("%s | %d | %s | %s | %s",
+                getTaskType(),
+                isDone ? 1 : 0,
+                this.description,
+                this.getDateString(start),
+                this.getDateString(end));
+    }
```
Is the indentation for this correct?

ivyy-poison commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,37 @@
+        this.isDone = isDone;
+    }
+
+    public void markTaskAsDone() {
+        this.isDone = true;
+    }
+
```
Perhaps look into creating short JavaDoc comments for public classes and methods to aid in readability?

ivyy-poison commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,188 @@
+        String description = parts[2];
+
+        switch (taskType) {
+            case "T":
+                addTask(new Todo(description, isDone));
+                break;
+
+            case "D":
+                String deadlineBy = parts[3];
+                addTask(new Deadline(description, deadlineBy, isDone));
+                break;
+
+            case "E":
```
Perhaps instead of using a singular letter to represent class types, use the entire string instead (eg: Todo, Deadline, and Event), to make the code more readable and scalable to potentially other forms of Tasks?

ivyy-poison reviewed (commented):
LGTM! Minor nitpicking but overall, your code is very neat and readable. Good job! :)

miljyy committed:
Create Command abstract class and its subclasses

miljyy committed:
Implement A-MoreOOP

miljyy committed:
Merge branch 'branch-MoreOOP'

miljyy committed:
Organise classes into commands package

miljyy committed:
Merge branch 'branch-Packages'

miljyy committed:
Merge branch 'add-gradle-support'

miljyy committed:
Set up gradle to run Buddy and Checkstyle

miljyy committed:
Add JUnit tests to test code behavior

miljyy committed:
Merge branch 'branch-JUnit'

miljyy committed:
Edit gradle to include jar

miljyy committed:
Add Javadoc to all classes except those in utils package.

miljyy committed:
Add Javadoc for classes in utils package

miljyy committed:
Merge branch 'branch-A-JavaDoc'

miljyy committed:
Edit code to comply with coding standard

miljyy committed:
Resolve conflicts in merge branch 'branch-A-CodingStandard'.

miljyy committed:
Fix bugs in code.

miljyy committed:
Implement Find command

miljyy committed:
Merge branch 'branch-Level-9'

miljyy committed:
Fix bugs

miljyy committed:
Add JavaDoc in Ui class

miljyy committed:
Add partial GUI to the chatbot

miljyy committed:
Edit commands to return strings

miljyy committed:
Make Java fxml attempt

miljyy committed:
Merge branch 'branch-Level-10'

miljyy committed:
Revert "Make Java fxml attempt"

This reverts commit 70179db0430f5a7710f3281ce73535aa24099fc7.

miljyy committed:
Merge branch 'branch-to-delete'

miljyy committed:
Delete command: make response line separated.

Response to delete command is in a single line.

A response with each statement starting in a new line is more
user-friendly because users face more difficulty in reading text that
runs in a single line.

miljyy committed:
Use assert to enforce Parser assumptions.

Parser currently does not enforce the input date string..

This runs the risk of having invalid input.

Adopt assert statements to ensure date is not empty and
AssertionError is thrown.

This makes it clearer why the input is invalid.

miljyy committed:
Use assert to enforce Storage assumptions

Storage currently does not enforce the filePath.

This runs the risk of having invalid input.

Adopt assert statements to ensure filePath is not empty and
AssertionError is thrown for missing keys.

This improves the visibility of configuration issues and makes it
clearer why exactly the chatbot failed.

miljyy committed:
Use assert to enforce TaskList assumptions

miljyy committed:
Parser class: implement A-CodingQuality
Parser class parse method is too long and have some complicated
expressions, affecting readability.

Let's break down the complicated expressions into steps and
multiple methods.

This improves code quality and readability.

miljyy committed:
Merge pull request #2 from miljyy/branch-Assertions

Implement assertions

miljyy committed:
Merge branch 'master' into branch-CodeQuality

miljyy committed:
Merge pull request #3 from miljyy/branch-CodeQuality

Parser class: implement A-CodingQuality

miljyy committed:
Find command: make matching case-insensitive
Find command is case-sensitive.

A case-insensitive find is more user-friendly because users cannot be
expected to remember the exact case of the keywords.

Let's update the search algorithm to use case-insensitive matching.

miljyy committed:
Merge branch 'branch-CodeQuality' of https://github.com/miljyy/ip into branch-CodeQuality

# Conflicts:
#	data/tasks.txt

miljyy committed:
Merge branch 'branch-CodeQuality'

miljyy committed:
Implement update function
Users cannot update tasks without deleting and re-adding.

Let's add an update function to allow users to update the task
description or date.

miljyy committed:
Merge branch 'branch-BCD-Extension'

miljyy committed:
Add a screenshot of the product

miljyy committed:
Update Ui.png

miljyy committed:
Add user guide

miljyy committed:
Minor fixes

miljyy committed:
Improve code quality and delete unused code

miljyy committed:
Update Gradle and Javadoc

miljyy committed:
Modify gradle and Buddy response after exit

miljyy committed:
Implement better GUI

miljyy committed:
Merge branch 'branch-A-BetterGUI'

miljyy committed:
Merge branch 'master' into branch-A-BetterGUI

miljyy committed:
Modify GUI application to close window after ExitCommand

miljyy committed:
Merge branch 'branch-A-BetterGUI'

miljyy committed:
Improve code to handle errors regarding invalid command formats

miljyy committed:
Merge branch 'branch-A-MoreErrorHandling'

miljyy committed:
Edit tests to fit updated methods to handle errors

miljyy committed:
Merge branch 'branch-A-MoreErrorHandling'

miljyy committed:
Implement testing for add task commands

miljyy committed:
Merge branch 'branch-A-MoreTesting'

miljyy committed:
Fix checkstyle violations

miljyy committed:
Update ReadMe user guide

miljyy committed:
Improve User interface and chatbot response

miljyy committed:
Make minor fixes

miljyy committed:
Upload image to user guide

miljyy committed:
Implement "help" command

miljyy committed:
Merge branch 'branch-A-BetterGUI'

miljyy committed:
Update README.md
